File: Data Structure Basics/Hashmaps/JavaScript Implementation/HashMapImplementation.js
function getCounter(arr)
{
    // initialize a hashmap to store each number and its count
    const counter = new Map();
    for (const num of arr)
    {
        // check if num is a key in the hash map
        if (counter.has(num))
        {
            // update the count of num to increase by 1
            counter.set(num, counter.get(num) + 1);
        }
        else
        {
            // set the count of num to 1
            counter.set(num,1);
        }
    }
    return counter;
}

function splitWords(s)
{
    return s == "" ? [] : s.split(' ');
}

function* main()
{
    const arr = splitWords(yield).map((v => parseInt(v)));
    const res = getCounter(arr)
    const sortedCounter = new Map([...res].sort((a,b) => a[0] - b[0]));
    for (const[index,value] of sortedCounter.entries())
    {
        console.log(index,value);
    }
}
class EOFError extends Error {}
{
    const gen = main();
    const next = (line) => gen.next(line).done && process.exit();
    let buf = '';
    next();
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', (data) => {
        const lines = (buf + data).split('\n');
        buf = lines.pop();
        lines.forEach(next);
    });
    process.stdin.on('end', () => {
        buf && next(buf);
        gen.throw(new EOFError());
    });
}

File: Data Structure Basics/Hashmaps/Java Implementation/Java Implementation/src/HashMapImplementation.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;
import java.util.stream.Collectors;
public class HashMapImplementation
{
    public static HashMap<Integer, Integer> getCounter(List<Integer> arr)
    {
        // initialize a hash map to store each number and its count
        HashMap<Integer, Integer> counter = new HashMap<>();
        for (int num : arr)
        {
            // check if num is  a key in the hah map
            if (counter.containsKey(num))
            {
                // update the count of num to increase by 1
                counter.put(num, counter.get(num) + 1);
            }
            else
            {
                // set the count of num to 1
                counter.put(num, 1);
            }
        }
        return counter;
    }

    public static List<String> splitWords(String s)
    {
        return s.isEmpty() ? List.of() : Arrays.asList(s.split(" "));
    }

    public static void main(String[] args)
    {
        Scanner scanner = new Scanner(System.in);
        List<Integer> arr = splitWords(scanner.nextLine()).stream().map(Integer::parseInt).collect(Collectors.toList());
        scanner.close();
        HashMap<Integer, Integer> counter = getCounter(arr);
        Map<Integer, Integer> sortedCounter = new TreeMap<>(counter);
        for (int key : sortedCounter.keySet()) {
            System.out.println(key + " " + sortedCounter.get(key));
        }
    }
}
